README.md
[![Build Status](https://secure.travis-ci.org/mikemaccana/agave.png?branch=master)](https://travis-ci.org/mikemaccana/agave)

Agave.js **safely** extends native JavaScript objects with helpful, intuitive methods that **make your code shorter and more readable**.

For full documentation, [visit agavejs.org](http://agavejs.org)



CHANGELOG.md
## 2.2.0 (2017-04-18)

 - es6-shim removed - now expects ES6. Current browsers or node version 6 and up.
 - `number..months` and `number..years` added.
 - Remove number methods where either Math.method an ES6 operator like `**` exists. Use those instead.

## 2.0.0 (2016-08-03)

 - agave now uses [semver](http://semver.org/). Since this is the first major breaking change we're releasing to 2.0.0
 - `agave.enable()` is now just `agave()` - ie the module exports a single function
 - number..seconds, number..minutes, number..hours etc are now attributes, rather than functions.
 - Add date.isOnWeekend(), date.withoutTime(), and date.clone()
 - agave itself is now written in ES6

## 0.6.1 (2015-02-22)

 - Replace array.contains with array.includes like ES7

## 0.6 (2015-02-22)

Features:

 - Use ES6, remove any function already in ES6 (@mikemaccana)
 - Remove requirejs support, just use CommonJS (@mikemaccana)
 - Remove DOM functions, these will soon appear in a separate library

## 0.4.4 (2014-05-29)

Features:

 - Function.repeat() & tests (@mikemaccana)
 - Enhance getPath() docs (@mikemaccana)

## 0.4.2 (2014-04-16)

Features:

 - Add element.getParentIndex(), object.compare() & tests (@mikemaccana)
 - Add number maths methods (@rich-harris)
 - Only allow agave to be enabled once per prefix (@rich-harris)
 - jshint now passes on tests (@mikemaccana)
 - Fix bug where getPath would run hasOwnProperty() on null
 - Remove unnecessary grunt dependency from master branch (only needed for gh-pages)

## 0.4.0 (2013-10-22)

Features:

  - Adds a prefixed global 'kind', which provides intuitive primitive/instance type checking (@mikemaccana, @notsorella)‎
  - Iteration function in object.forEach() now given value as second argument (@alexgraul)‎
  - Allow new properties to be overwritable if users want (@rich-harris)
  - 'grunt watch' support in gh-pages, for rebuilding css whenever less changes

Bugfixes:

  - Doc typos (#1, @nubs) (#2, @evanhahn)

## 0.2.0 (2013-06-14)

 - Initial release


test/tests.js
// Tests. Mocha/assert style. See
// http://visionmedia.github.com/mocha/
// http://nodejs.org/docs/latest/api/assert.html

/*jshint multistr:true */
/*jshint -W010 */
/*jshint -W053 */
/*jshint -W054 */
// 10 allows us to test Object
// 53 allows us to test Number, String etc constructors.
// 54 allows us to test Function constructors.

var assert = require('assert');
var agave = require('../index.js');

var mockObject = {
	foo: 'bar',
	baz: {
		bam:'boo',
		zar:{
			zog:'victory'
		}
	},
	null:{
		'yarr':{
			'parrot':'ahoy'
		}
	}
};

agave('av');

suite('Array.includes', function(){
	test('fetches the item accurately', function(){
		assert(['one','two','three'].avincludes('two') );
	});
	test('handles missing items accurately', function(){
		assert( ! ['one','two','three'].avincludes('notthere') );
	});
});

suite('Array.extend', function(){
	test('extends the array accurately', function(){
		assert.deepEqual([1,2,3].avextend([4,5]), [1,2,3,4,5] );
	});
});

suite('String.reverse', function(){
	test('reverses strings accurately', function(){
		assert.equal('Hello world'.avreverse(), 'dlrow olleH');
	});
});

suite('String.leftStrip', function(){
	test('strips from the left accurately', function(){
		assert.equal('Hello world'.avleftStrip('Hle'), 'o world');
	});
});

suite('String.rightStrip', function(){
	test('strips from the right accurately', function(){
		assert.equal('Hello world'.avrightStrip('ldr'), 'Hello wo');
	});
});

suite('String.rightStrip', function(){
	test('strips from the left accurately with a single character', function(){
		assert.equal('a'.avleftStrip('a'), '');
	});
});

suite('String.strip', function(){
	test('strips from the both sides accurately', function(){
		assert.equal('Hello world'.avstrip('Hld'), 'ello wor');
	});
});

suite('Object.getKeys', function(){
	test('fetches keys accurately', function(){
		assert.deepEqual(mockObject.avgetKeys(), ["foo","baz","null"] );
	});
});

suite('Object.getSize', function(){
	test('counts keys accurately', function(){
		assert.equal(mockObject.avgetSize(), 3);
	});
});

suite('Array.findItem', function(){
	test('correctly finds items that match the function', function(){
		assert.equal(['one','two','three'].avfindItem(function(item){
			return (item === 'three');
		}), 'three');
	});
});

suite('Array.remove', function () {
	test('correctly removes the given member', function (){
		var arr = [1,2,3,4,5];
		arr.avremove(3);
		assert.deepEqual(arr, [1,2,4,5]);
	});
	if('returns true if the given member was in the array', function () {
		assert.equal([1,2,3,4,5].remove(3), true);
	});
	if('returns false if the given member was not in the array', function () {
		assert.equal([1,2,3,4,5].remove(6), false);
	});
});

suite('Object.getPath', function(){
	test('returns undefined when a value is missing', function(){
		assert.equal(mockObject.avgetPath(['foo','pineapple']), undefined);
	});
	test('returns the value when the provided keys exist', function(){
		assert.equal(mockObject.avgetPath(['baz','zar','zog']), 'victory');
	});
	test('returns the value when the provided keys exist, even if null is on the path', function(){
		assert.equal(mockObject.avgetPath([null,'yarr','parrot']), 'ahoy');
	});
	test('works using Unix-style paths', function(){
		assert.equal(mockObject.avgetPath('/baz/zar/zog'), 'victory');
	});
});

suite('Object.clone', function(){
	var copyObject = mockObject.avclone();
	test('clones objects so that modification to the new object will not affect the original', function(){
		copyObject.baz.bam = 'newvalue';
		assert.equal(copyObject.avgetPath(['baz','bam']), 'newvalue');
		assert.equal(mockObject.avgetPath(['baz','bam']), 'boo');
	});
});

suite('Object.forEach', function(){
	var keyResults = [];
	var valueResults = [];
	mockObject.avforEach(function(key, value){
		keyResults.push(key);
		valueResults.push(value);
	});
	test('iterates over keys properly', function(){
		assert.deepEqual(keyResults, ["foo","baz","null"]);
	});
	test('iterates over values properly', function(){
		assert.deepEqual(valueResults, [
			"bar",
			{"bam":"boo",
				"zar":{
					"zog":"victory"
				}
			},
			{"yarr":
				{"parrot":"ahoy"}
			}
		]);
	});
});

suite('Object.extend', function(){
	var results = mockObject.avclone().avextend({
		'gnar':{
			shub:'zoo'
		},
		'gert':{
			yaz:'frub'
		}
	});

	test('creates extends the object with the new properties', function(){
		assert.deepEqual(results, {
			"foo":"bar",
			"baz":{
				"bam":"boo",
				"zar":{
					"zog":"victory"}
				},
				"null":{
					"yarr":{
						"parrot":"ahoy"
					}
				},
			"gnar":{
				"shub":"zoo"
			},
			"gert":{
				"yaz":"frub"
			}
		});
	});
});

suite('Object.compare', function(){
	test('accurately identifies similar objects', function(){
		var identicalObject = {
			foo: 'bar',
			baz: {
				bam:'boo',
				zar:{
					zog:'victory'
				}
			},
			null:{
				'yarr':{
					'parrot':'ahoy'
				}
			}
		};
		assert(mockObject.avcompare(identicalObject));
	});
	test('accurately identifies different objects', function(){
		var differentObject = {
			foo: 'bar',
			baz: {
				bam:'boo',
				zar:{
					zog:'victory'
				}
			}
		};
		assert.equal(mockObject.avcompare(differentObject), false);
	});
});

suite('Function.throttle', function(){
	var valueShouldOnlyBeOne = 0; // Since the function should only run once
	var timesFunctionHasRan = 0;
	var maxTimesToRun = 3;
	test('stops function calls overlapping', function(done){
		var intervalID = setInterval(function(){
			// thottledFunction would normally run three times in this loop, 50ms apart, but .avthrottle() means it
			// will only be run once after 70ms of inactivity
			var thottledFunction = function(){
				valueShouldOnlyBeOne++;
			}.avthrottle(70);
			thottledFunction();
			timesFunctionHasRan++;
			if ( timesFunctionHasRan === maxTimesToRun ) {
				clearInterval(intervalID);
				assert.equal(valueShouldOnlyBeOne, 1);
				done();
			}
		}, 50);
	});
});

suite('Function.repeat', function(){

	test('repeats', function(done){
		var count = 0
		var increment = function(){
			count += 1
			if ( count === 3 ) {
				done();
			}
		}
		increment.repeat([], 50, true)
	});
	test('repeats with arguments omitted', function(done){
		var count = 0
		this.timeout(11 * 1000);
		var increment = function(){
			count += 1
			if ( count === 3 ) {
				done();
			}
		}
		increment.repeat(50, true)
	});
});



suite('Number.days', function(){
	test('correctly converts a number to days in seconds', function(){
		assert.equal((5).avdays, 432000000);
	});
});

suite('Number.weeks.before and .after', function(){
	test('correctly converts a number to a period in weeks before a set date', function(){
		var someDate = new Date('Thu Jun 06 2013 22:44:05 GMT+0100 (UTC)');
		var timezoneOffset = someDate.getTimezoneOffset();
		var targetDate = new Date('Thu May 16 2013 22:44:05 GMT+0100 (UTC)')
		assert.equal((3).avweeks.avbefore(someDate).getDate(), targetDate.getDate());
	});
	test('correctly converts a number to a period in weeks after a set date', function(){
		var someDate = new Date('Thu Jun 27 2013 22:44:05 GMT+0100 (UTC)');
		var timezoneOffset = someDate.getTimezoneOffset();
		var targetDate = new Date('Thu Jun 06 2013 22:44:05 GMT+0100 (UTC)');
		assert.equal((3).avweeks.avbefore(someDate).getDate(), targetDate.getDate());
	});
});

suite('Date functions', function(){
	test('isOnWeekend works', function(){
		var dayOnWeekend = new Date('Sun Jul 31 2016 18:55:19 GMT+0100 (BST)');
		assert(dayOnWeekend.isOnWeekend())
	});

	test('isOnWeekend does not give false positives', function(){
		var dayOnWeek = new Date('Wed Aug 03 2016 18:55:19 GMT+0100 (BST)')
		assert.equal(dayOnWeek.isOnWeekend(), false)
	});

	test('withoutTime returns a time at midnight', function(){
		var someDay = new Date('Sun Jul 31 2016 18:55:19 GMT+0100 (BST)')
		assert.deepEqual(someDay.withoutTime(), new Date('Sun Jul 31 2016 00:00:00 GMT+0100 (BST)'))
	});

	test('date clone makes new dates which are not affected by changes in original', function(){
		var someDay = new Date('Sun Jul 31 2016 18:55:19 GMT+0100 (BST)')
		var someDayCopy = someDay.clone();
		someDay.setFullYear(2006)
		assert.deepEqual(someDayCopy, new Date('Sun Jul 31 2016 18:55:19 GMT+0100 (BST)'))
	});
});

suite('Agave doesn\'t affect for loops', function(){
	it ('doesn\'t. really', function(){
		for ( var key in mockObject ) {
			assert( ! ['avgetKeys','avgetSize','avgetPath'].avincludes(key) );
		}
	});
});

suite('kind', function(){
	test('shows number-like things as numbers', function(){
		assert(avkind(37) === 'Number');
		assert(avkind(3.14) === 'Number');
		assert(avkind(Math.LN2) === 'Number');
		assert(avkind(Infinity) === 'Number');
		assert(avkind(Number(1)) === 'Number');
		assert(avkind(new Number(1)) === 'Number');
	});
	test('shows NaN as NaN', function(){
		assert(avkind(NaN) === 'NaN');
	});
	test('Shows strings as strings', function(){
		assert(avkind('') === 'String');
		assert(avkind('bla') === 'String');
		assert(avkind(String("abc")) === 'String');
		assert(avkind(new String("abc")) === 'String');
	});
	test('shows strings accurately', function(){
		assert(avkind(true) === 'Boolean');
		assert(avkind(false) === 'Boolean');
		assert(avkind(new Boolean(true)) === 'Boolean');
	});
	test('shows arrays accurately', function(){
		assert(avkind([1, 2, 4]) === 'Array');
		assert(avkind(new Array(1, 2, 3)) === 'Array');
	});
	test('shows objects accurately', function(){
		assert(avkind({a:1}) === 'Object');
		assert(avkind(new Object()) === 'Object');
	});
	test('shows dates accurately', function(){
		assert(avkind(new Date()) === 'Date');
	});
	test('loves Functions too', function(){
		assert(avkind(function(){}) === 'Function');
		assert(avkind(new Function("console.log(arguments)")) === 'Function');
		assert(avkind(Math.sin) === 'Function');
	});
	test('shows undefined accurately', function(){
		assert(avkind(undefined) === 'undefined');
	});
	test('shows null accurately', function(){
		assert(avkind(null) === 'null');
	});
});

suite('Functions work with no prefix at all', function(){
	agave();
	test('strips from the right accurately', function(){
		assert.equal('Hello world'.rightStrip('ldr'), 'Hello wo');
	});
});





gulpfile.js
var gulp = require('gulp'),
  prefixer = require('gulp-autoprefixer'),
  sass = require('gulp-sass'),
  del = require('del'),
  browserify = require('gulp-browserify')

var path = require('path');

gulp.task('sass', function () {
  gulp
    .src('./scss/style.scss')
    .pipe(sass({
      paths: ['scss']
    }))
    .pipe(prefixer('last 2 versions', 'ie 9'))
    .pipe(gulp.dest('./css'));
});

// Delete old JS files
gulp.task('clean', function(done) {
  del(['./js/dist'], done);
});

// Browserify our code
gulp.task('js', ['clean'], function() {
  // Browserify/bundle the JS.
  gulp.src('./js/src/index.js')
    .pipe(browserify({
      insertGlobals : true,
      debug : ! gulp.env.production
    }))
    .pipe(gulp.dest('./js/dist'))
});



// The default task (called when you run `gulp`)
gulp.task('default', ['sass', 'js'], function() {

  // Watch files and run tasks if they change
  gulp.watch('./scss/*.scss', ['sass']);
  gulp.watch('./js/src/*.js', ['js']);

});



js/agave.js
// Agave.JS
// I'm a UMD module (works in RequireJS and CommonJS-like environments)
// See https://github.com/umdjs
(function (root, factory) {
  if (typeof exports === 'object') {
    // Node. Does not work with strict CommonJS, but
    // only CommonJS-like enviroments that support module.exports,
    // like Node.
    module.exports = factory();
  } else if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(factory);
  } else {
    // Browser globals (root is window)
    root.returnExports = factory();
  }
}(this, function () {

  // Extend objects with Agave methods, using
  var enable = function(prefix){
    var global = this;

    var SECONDS = 1000;
    var MINUTES = 60 * SECONDS;
    var HOURS = 60 * MINUTES;
    var DAYS = 24 * HOURS;
    var WEEKS = 7 * DAYS;

    // object.getKeys() returns an array of keys
    var getKeys = function(){
      return Object.keys(this);
    };

    // object.getSize() returns the number of properties in the object
    var getSize = function() {
      return Object.keys(this).length;
    };

    // string.reverse()
    var reverse = function() {
      return this.split("").reverse().join("");
    };

    // string.leftStrip(stripChars) returns the string with the leading chars removed
    var leftStrip = function(stripChars) {
      var result = this;
      while ( true ) {
        // Note result could be zero characters
        if ( ! stripChars[prefix+'contains'](result.charAt(0)) || ! result) {
          return result;
        } else {
          result = result.slice(1);
        }
      }
    };

    // string.rightStrip(stripChars) returns the string with the trailing chars removed
    var rightStrip = function(stripChars) {
      return this[prefix+'reverse']()[prefix+'leftStrip'](stripChars)[prefix+'reverse']();
    };

    // string.strip(stripChars) returns the string with the leading and trailing chars removed
    var strip = function(stripChars) {
      return this[prefix+'leftStrip'](stripChars)[prefix+'rightStrip'](stripChars);
    };

    // object.getPath - get the value of the nested keys provided in the object.
    // If any are missing, return undefined. Used for checking JSON results.
    var getPath = function(pathItems) {
      var obj = this;
      var delim = '/';
      var result;
      var still_checking = true;
      // Handle Unix style paths
      if ( typeof(pathItems) === 'string' ) {
        pathItems = pathItems[prefix+'strip'](delim).split(delim);
      }
      pathItems.forEach( function(pathItem) {
        if ( still_checking ) {
          if ( ! obj.hasOwnProperty(pathItem) ) {
            result = undefined;
            still_checking = false;
          } else {
            result = obj[pathItem];
          }
          obj = obj[pathItem];
        }
      });
      return result;
    };

    // array.findItem(test_function) returns the first item that matches the test_function
    var findItem = function(test_function){
      var arr = this;
      var last_index;
      var found = arr.some(function(item, index) {
        last_index = index;
        return test_function(item);
      });
      if ( found ) {
        return arr[last_index];
      } else {
        return null;
      }
    };

    // string.endsWith(suffix) returns true if string ends with the suffix
    var endsWith = function(suffix) {
      return this.indexOf(suffix, this.length - suffix.length) !== -1;
    };

    // string.endsWith(prefix) returns true if string ends with the prefix
    var startsWith = function(prefix){
      return this.slice(0, prefix.length) === prefix;
    };

    // array.contains(item) returns true if an array contains an item
    // string.contains(substring) returns true if a string contains a substring
    var contains = function(item){
      return ( this.indexOf(item) !== -1);
    };

    // Extend an array with another array.
    // Cleverness alert: since .apply() accepts an array of args, we use the new_array as all the args to push()
    var extend = function(new_array) {
      Array.prototype.push.apply(this, new_array);
      return this;
    };

    // string.repeat() repeat a string 'times' times. Borrowed from ES6 shim at https://github.com/paulmillr/es6-shim
    var repeat = function(times) {
      if (times < 1) return '';
      if (times % 2) return this[prefix+'repeat'](times - 1) + this;
      var half = this[prefix+'repeat'](times / 2);
      return half + half;
    };

    // Clone an object recursively
    var clone = function() {
      var newObj = (this instanceof Array) ? [] : {};
      for (var key in this) {
        if (this[key] && typeof this[key] == "object") {
          newObj[key] = this[key][prefix+'clone']();
        } else {
          newObj[key] = this[key];
        }
      }
      return newObj;
    };

    var arrayClone = function(){return this.slice();};

    // Array toNodeList converts arrays to NodeLists
    var toNodeList = function(){
      var fragment = document.createDocumentFragment();
      this.forEach(function(item){
        fragment.appendChild(item);
      });
      return fragment.childNodes;
    };

    // Convert Number to (function name). +ensures type returned is still Number
    var seconds = function() {
      return +this * SECONDS;
    };
    var minutes = function() {
      return +this * MINUTES;
    };
    var hours = function() {
      return +this * HOURS;
    };
    var days = function() {
      return +this * DAYS;
    };
    var weeks = function() {
      return +this * WEEKS;
    };

    // Helper function for before() and after()
    var getTimeOrNow = function(date) {
      return (date || new Date()).getTime()
    }

    // Return Number of seconds to time delta from date (or now if not specified)
    var before = function(date) {
      var time = getTimeOrNow(date);
      return new Date(time-(+this));
    }

    // Return Number of seconds to time delta after date (or now if not specified)
    var after = function(date) {
      var time = getTimeOrNow(date);
      return new Date(time+(+this));
    }

    // Add a new element as a child of this element
    var createChild = function(name, attributes, text) {
      var newElement = document.createElement(name);
      if ( attributes ) {
        for (var attribute in attributes) {
          newElement.setAttribute(attribute, attributes[attribute]);
        }
      }
      if ( text ) {
        newElement.textContent = text;
      }
      return this.appendChild(newElement);
    };

    // Apply the CSS styles
    var applyStyles = function(styles) {
      for ( var style in styles ) {
        this.style[style] = styles[style];
      }
      return this;
    };

    // Toggle a class
    var toggleClass = function(className) {
      if ( this.classList.contains(className) ) {
        this.classList.remove(className);
      } else {
        this.classList.add(className);
      }
      return this;
    };

    // Return nodeList of an elements parent elements from closest to farthest
    var ancestorNodes = function(selector) {
      var ancestors = [];
      var parent = this.parentNode;
      // While parents are 'element' type nodes
      // See https://developer.mozilla.org/en-US/docs/DOM/Node.nodeType
      while ( parent && parent.nodeType && parent.nodeType === 1 ) {
        if ( selector ) {
          if ( parent.matches(selector) ) {
            ancestors.append(parent);
          }
        } else {
          ancestors.append(parent);
        }
        parent = parent.parentNode;
      }
      // Return a NodeList to be consistent with childNodes
      return ancestors.toNodeList();
    };

    // Polyfill if Element.prototype.matches doesn't exist.
    var prefixedMatchesMethod = ( !this.Element || Element.prototype.msMatchesSelector || Element.prototype.mozMatchesSelector || Element.prototype.webkitMatchesSelector || Element.prototype.oMatchesSelector);

    // Add method as a non-enumerable property on obj with the name methodName
    var addMethod = function( global, objectName, prefix, methodName, method) {
      var objectToExtend = global[objectName];
      methodName = prefix ? prefix+methodName: methodName;
      // Check - NodeLists and Elements don't always exist on all JS implementations
      if ( objectToExtend ) {
        // Don't add if the method already exists
        if ( ! objectToExtend.prototype.hasOwnProperty(methodName) ) {
          Object.defineProperty( objectToExtend.prototype, methodName, {
            value: method,
            enumerable: false
          });
        }
      }
    };

    var newMethods = {
      'Array':{
        'findItem':findItem,
        'extend':extend,
        'contains':contains,
        'clone':arrayClone,
        'toNodeList':toNodeList
      },
      'Object':{
        'getKeys':getKeys,
        'findItem':findItem,
        'getSize':getSize,
        'getPath':getPath,
        'contains':contains,
        'clone':clone
      },
      'String':{
        'endsWith':endsWith,
        'startsWith':startsWith,
        'repeat':repeat,
        'reverse':reverse,
        'leftStrip':leftStrip,
        'rightStrip':rightStrip,
        'strip':strip,
        'contains':contains,
        'forEach':Array.prototype.forEach // Strings and NodeLists don't have .forEach() standard but the one from Array works fine
      },
      'Number':{
        'seconds':seconds,
        'minutes':minutes,
        'hours':hours,
        'days':days,
        'weeks':weeks,
        'before':before,
        'after':after
      },
      'Element':{
        'createChild':createChild,
        'ancestorNodes':ancestorNodes,
        'matches':prefixedMatchesMethod,
        'applyStyles':applyStyles,
        'toggleClass':toggleClass
      },
      'NodeList':{
        'forEach':Array.prototype.forEach,
        'reverse':Array.prototype.reverse
      }
    };
    for ( var objectName in newMethods ) {
      for ( var methodName in newMethods[objectName] ) {
        addMethod(global, objectName, prefix, methodName, newMethods[objectName][methodName]);
      }
    }
  }.bind();

  // Just return a value to define the module export.
  // This example returns an object, but the module
  // can return a function as the exported value.
  return {
    enable:enable
  };
}));

index.js
var enabledPrefixes = {}; // Only allow agave to be enabled once per prefix

// Extend objects with Agave methods, using the prefix provided.
var enable = function(prefix){
	"use strict";

	prefix = prefix || '';

	if ( enabledPrefixes[prefix] ) {
		return;
	}

	const MILLISECONDS_IN_SECOND = 1000,
		SECONDS_IN_MINUTE = 60,
		MINUTES_IN_HOUR = 60,
		HOURS_IN_DAY = 24,
		DAYS_IN_WEEK = 7,
		DAYS_IN_MONTH = 30,
		DAYS_IN_YEAR = 365;

	const SUNDAY = 0,
		MONDAY = 1,
		TUESDAY = 2,
		WEDNESDAY = 3,
		THURSDAY = 4,
		FRIDAY = 5,
		SATURDAY = 6;

	// object.getKeys() returns an array of keys
	var getKeys = function(){
		return Object.keys(this);
	};

	// object.getSize() returns the number of properties in the object
	var getSize = function() {
		return Object.keys(this).length;
	};

	// string.reverse()
	var reverse = function() {
		return this.split("").reverse().join("");
	};

	// string.leftStrip(stripChars) returns the string with the leading chars removed
	var leftStrip = function(stripChars) {
		var result = this;
		while ( true ) {
			// Note result could be zero characters
			if ( ! stripChars.includes(result.charAt(0)) || ! result) {
				return result;
			} else {
				result = result.slice(1);
			}
		}
	};

	// string.rightStrip(stripChars) returns the string with the trailing chars removed
	var rightStrip = function(stripChars) {
		return this[prefix+'reverse']()[prefix+'leftStrip'](stripChars)[prefix+'reverse']();
	};

	// string.strip(stripChars) returns the string with the leading and trailing chars removed
	var strip = function(stripChars) {
		return this[prefix+'leftStrip'](stripChars)[prefix+'rightStrip'](stripChars);
	};

	// object.getPath - get the value of the nested keys provided in the object.
	// If any are missing, return undefined. Used for checking JSON results.
	var getPath = function(pathItems) {
		var currentObject = this;
		var delim = '/';
		var result;
		var stillChecking = true;
		// Handle Unix style paths
		if ( typeof(pathItems) === 'string' ) {
			pathItems = pathItems[prefix+'strip'](delim).split(delim);
		}
		pathItems.forEach( function(pathItem) {
			if ( stillChecking ) {
				if ( ( currentObject === null ) || ( ! currentObject.hasOwnProperty(pathItem) ) ) {
					result = undefined;
					stillChecking = false;
				} else {
					result = currentObject[pathItem];
					currentObject = currentObject[pathItem];
				}
			}
		});
		return result;
	};

	// object.extent(object) adds the keys/values from the newObject provided
	var objectExtend = function(newObject) {
		for ( var key in newObject ) {
			this[key] = newObject[key];
		}
		return this;
	};

	// array.findItem(testFunction) returns the first item that matches the testFunction
	var findItem = function(testFunction){
		var lastIndex;
		var found = this.some(function(item, index) {
			lastIndex = index;
			return testFunction(item);
		});
		if ( found ) {
			return this[lastIndex];
		} else {
			return null;
		}
	};

	// Run after it hasn't been invoked for 'wait' ms.
	// Useful to stop repeated calls to a function overlapping each other (sometimes called 'bouncing')
	var throttle = function(wait, immediate) {
		var timeoutID;
		var originalFunction = this;
		return function() {
			var context = this;
			var delayedFunction = function() {
				timeoutID = null;
				if ( ! immediate ) {
					originalFunction.apply(context, arguments);
				}
			};
			var callNow = immediate && ! timeoutID;
			clearTimeout(timeoutID);
			timeoutID = setTimeout(delayedFunction, wait);
			if (callNow) {
				originalFunction.apply(context, arguments);
			}
		};
	};

	// Run repeatedly
	var functionRepeat = function(first, second, third){
		var args, interval, leadingEdge;
		if ( arguments.length === 2 ) {
			args = [];
			interval = first;
			leadingEdge = second;
		} else {
			args = first;
			interval = second;
			leadingEdge = third;
		}
		if ( leadingEdge ) {
			this.apply(null, args);
		}
		return setInterval(function(){
			this.apply(null, args);
		}.bind(this), interval);
	};

	// Extend an array with another array.
	// Cleverness alert: since .apply() accepts an array of args, we use the newArray as all the args to push()
	var arrayExtend = function(newArray) {
		Array.prototype.push.apply(this, newArray);
		return this;
	};

	// string.toHash() return a hashed value of a string
	// From http://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript-jquery
	var toHash = function(){
		var hash = 0,
			length = this.length,
			char;
		if ( ! length ) {
			return hash;
		}
		for (var index = 0; index < length; index++) {
			char = this.charCodeAt(index);
			hash = ((hash<<5)-hash)+char;
			hash |= 0; // Convert to 32bit integer
		}
		return hash;
	};

	// Clone an object recursively
	var clone = function() {
		var newObj = (this instanceof Array) ? [] : {};
		for (var key in this) {
			if (this[key] && typeof this[key] == "object") {
				newObj[key] = this[key][prefix+'clone']();
			} else {
				newObj[key] = this[key];
			}
		}
		return newObj;
	};

	// compare an object with another object
	var compare = function(otherObject){
		var hashObject = function(object){
			return JSON.stringify(object)[prefix+'toHash']();
		};
		return ( hashObject(this) === hashObject(otherObject) );
	};

	// Iterate over an objects keys
	// Unlike a regular for ( var key in object )
	// an additional scope is created, which avoids last-item looping probs
	var objectForEach = function(callback){
		for ( var key in this ) {
			callback(key, this[key]);
		}
	};

	var arrayClone = function(){
		return this.slice();
	};

	// Array remove removes an item from an array, if it exists
	var arrayRemove = function(member){
		var index = this.indexOf(member);
		if (index !== -1 ) {
			this.splice(index, 1);
			return true;
		}
		return false;
	};

	var arrayFirst= function(count){
		if ( ! count ) {
			return this[0];
		} else {
			return this.slice(Math.max(arr.length - count, 1));
		}
	};

	var arrayLast = function(count){
		if ( ! count ) {
			return this[this.length - 1];
		} else {
			return this.slice(Math.max(this.length - count, 1));
		}
	};

	// Helper function for before() and after()
	var getTimeOrNow = function(date) {
		return (date || new Date()).getTime();
	};

	// Return Number of seconds to time delta from date (or now if not specified)
	var before = function(date) {
		var time = getTimeOrNow(date);
		return new Date(time-(+this));
	};

	// Return Number of seconds to time delta after date (or now if not specified)
	var after = function(date) {
		var time = getTimeOrNow(date);
		return new Date(time+(+this));
	};

	var toSeconds = function() {
		return this * MILLISECONDS_IN_SECOND;
	}

	var toMinutes = function() {
		return this.seconds * SECONDS_IN_MINUTE;
	}

	var toHours = function() {
		return this.minutes * MINUTES_IN_HOUR;
	}

	var toDays = function() {
		return this.hours * HOURS_IN_DAY;
	}

	var toWeeks = function() {
		return this.days * DAYS_IN_WEEK;
	}

	var toMonths = function() {
		return this.days * DAYS_IN_MONTH;
	}

	var toYears = function() {
		return this.days * DAYS_IN_YEAR;
	}

	var isOnWeekend = function(){
		return this.getDay() === SUNDAY || this.getDay() === SATURDAY
	}

	var withoutTime = function(){
		var copy = new Date(this)
		copy.setHours(0, 0, 0, 0, 0)
		return copy;
	}

	var dateClone = function(){
		return new Date(this.getTime())
	}

	var kind = function(item) {
		var getPrototype = function(item) {
			return Object.prototype.toString.call(item).slice(8, -1);
		};
		var kind, Undefined;
		if (item === null ) {
			kind = 'null';
		} else {
			if ( item === Undefined ) {
				kind = 'undefined';
			} else {
				var prototype = getPrototype(item);
				if ( ( prototype === 'Number' ) && isNaN(item) ) {
					kind = 'NaN';
				} else {
					kind = prototype;
				}
			}
		}
		return kind;
	};

	// Polyfill if Element.prototype.matches doesn't exist.
	var prefixedMatchesMethod = ( ! global.Element || Element.prototype.msMatchesSelector || Element.prototype.mozMatchesSelector || Element.prototype.webkitMatchesSelector || Element.prototype.oMatchesSelector);

	// Add method as a non-enumerable property on obj with the name methodName
	var addMethod = function( global, objectName, prefix, methodName, method) {
		var objectToExtend = global[objectName];
		methodName = prefix ? prefix+methodName: methodName;
		// Check - NodeLists and Elements don't always exist on all JS implementations
		if ( objectToExtend ) {
			// Don't add if the method already exists
			if ( ! objectToExtend.prototype.hasOwnProperty(methodName) ) {
				Object.defineProperty( objectToExtend.prototype, methodName, {
					value: method,
					writable: true
				});
			}
		}
	};

	// There's not always a 1:1 match of functions to method names. Eg, some objects share methods,
	// others re-use inbuilt methods from other objects.
	var newMethods = {
		'Array':{
			'findItem':findItem,
			'extend':arrayExtend,
			'includes': String.prototype.includes,
			'clone':arrayClone,
			'remove':arrayRemove,
			'first':arrayFirst,
			'last':arrayLast
		},
		'Object':{
			'getKeys':getKeys,
			'getSize':getSize,
			'getPath':getPath,
			'clone':clone,
			'forEach':objectForEach,
			'extend':objectExtend,
			'compare':compare
		},
		'String':{
			'reverse':reverse,
			'leftStrip':leftStrip,
			'rightStrip':rightStrip,
			'strip':strip,
			'toHash':toHash,
			'forEach':Array.prototype.forEach // Strings and NodeLists don't have .forEach() standard but the one from Array works fine
		},
		'Function':{
			'throttle':throttle,
			'repeat':functionRepeat
		},
		'Number':{
			'before':before,
			'after':after
		},
		'Date':{
			'isOnWeekend': isOnWeekend,
			'withoutTime': withoutTime,
			'clone': dateClone
		}
	};

	for ( var objectName in newMethods ) {
		for ( var methodName in newMethods[objectName] ) {
			addMethod(global, objectName, prefix, methodName, newMethods[objectName][methodName]);
		}
	}

	// Add sttribute as a non-enumerable property on obj with the name methodName
	var addNewAttribute = function( global, objectName, prefix, methodName, method) {
		var objectToExtend = global[objectName];
		methodName = prefix ? prefix+methodName: methodName;
		// Check - NodeLists and Elements don't always exist on all JS implementations
		if ( objectToExtend ) {
			// Don't add if the method already exists
			if ( ! objectToExtend.prototype.hasOwnProperty(methodName) ) {
				Object.defineProperty( objectToExtend.prototype, methodName, {
					get: method
				});
			}
		}
	};

	var addTimeExtension = function(name, getterFunction){
		Object.defineProperty(Number.prototype, name, {
			get: getterFunction
		})
	}

	var newAttributes = {
		'Number':{
			'second':toSeconds,
			'seconds':toSeconds,
			'minute':toMinutes,
			'minutes':toMinutes,
			'hour':toHours,
			'hours':toHours,
			'day':toDays,
			'days':toDays,
			'week':toWeeks,
			'weeks':toWeeks,
			'month':toMonths,
			'months':toMonths,
			'year':toYears,
			'years':toYears
		}
	}

	for ( var objectToGetNewAttribute in newAttributes ) {
		for ( var attributeName in newAttributes[objectToGetNewAttribute] ) {
			addNewAttribute(global, objectToGetNewAttribute, prefix, attributeName, newAttributes[objectToGetNewAttribute][attributeName]);
		}
	}

	// Add a function to the global
	var addGlobal = function( global, globalName, prefix, globalFunction) {
		globalName = prefix ? prefix+globalName: globalName;
		// Don't add if the global already exists
		if ( ! global.hasOwnProperty(globalName) ) {
			global[globalName] = globalFunction;
		}
	};
	addGlobal(global, 'kind', prefix, kind);

	enabledPrefixes[prefix] = true;
}.bind();

module.exports = enable;
